packages/server/modules/workspaces/rest/sso.ts

```diff
@@ -11,272 +15,586 @@ import {
+import {
+  countAdminUsersFactory,
+  getUserFactory,
+  legacyGetUserFactory,
+  storeUserAclFactory,
+  storeUserFactory,
+  UserWithOptionalRole
+} from '@/modules/core/repositories/users'
 import {
   finalizeAuthMiddlewareFactory,
+  moveAuthParamsToSessionMiddlewareFactory,
   sessionMiddlewareFactory
 } from '@/modules/auth/middleware'
+import {
+  deleteInviteFactory,
+  deleteServerOnlyInvitesFactory,
+  findInviteFactory,
+  updateAllInviteTargetsFactory
+} from '@/modules/serverinvites/repositories/serverInvites'
+import { createUserFactory } from '@/modules/core/services/users/management'
+import { getServerInfoFactory } from '@/modules/core/repositories/server'
+import { validateAndCreateUserEmailFactory } from '@/modules/core/services/userEmails'
+import { UsersEmitter } from '@/modules/core/events/usersEmitter'
+import { finalizeInvitedServerRegistrationFactory } from '@/modules/serverinvites/services/processing'
+import { requestNewEmailVerificationFactory } from '@/modules/emails/services/verification/request'
+import { deleteOldAndInsertNewVerificationFactory } from '@/modules/emails/repositories'
+import { sendEmail } from '@/modules/emails/services/sending'
+import { renderEmail } from '@/modules/emails/services/emailRendering'
 import { createAuthorizationCodeFactory } from '@/modules/auth/repositories/apps'
-import { legacyGetUserFactory } from '@/modules/core/repositories/users'
-
-const router = Router()
+import { getDefaultSsoSessionExpirationDate } from '@/modules/workspaces/domain/sso/logic'
+import { GetWorkspaceBySlug } from '@/modules/workspaces/domain/operations'
+import {
+  GetWorkspaceSsoProvider,
+  UpsertUserSsoSession
+} from '@/modules/workspaces/domain/sso/operations'
+import { GetUser } from '@/modules/core/domain/users/operations'
+import { FindEmail } from '@/modules/core/domain/userEmails/operations'
+import { AuthorizeResolver } from '@/modules/shared/domain/operations'
+import { authorizeResolverFactory } from '@/modules/shared/services/auth'
+import { getRolesFactory } from '@/modules/shared/repositories/roles'
+import {
+  getUserAclRoleFactory,
+  getUserServerRoleFactory
+} from '@/modules/shared/repositories/acl'
+import { getStreamFactory } from '@/modules/core/repositories/streams'
+import {
+  buildAuthRedirectUrl,
+  buildErrorUrl,
+  buildFinalizeUrl,
+  decryptorFactory,
+  encryptorFactory,
+  parseCodeVerifier
+} from '@/modules/workspaces/helpers/sso'
 
+const moveAuthParamsToSessionMiddleware = moveAuthParamsToSessionMiddlewareFactory()
 const sessionMiddleware = sessionMiddlewareFactory()
 const finalizeAuthMiddleware = finalizeAuthMiddlewareFactory({
   createAuthorizationCode: createAuthorizationCodeFactory({ db }),
   getUser: legacyGetUserFactory({ db })
 })
 
-const buildAuthRedirectUrl = (workspaceSlug: string): URL =>
-  new URL(
-    `/api/v1/workspaces/${workspaceSlug}/sso/oidc/callback?validate=true`,
-    getServerOrigin()
+export default (app: Application) => {
```

Comment:
i've been using an express router creation, rather than asking for an app instance here, its just better IOC. See the gatekeeper rest router. https://github.com/specklesystems/speckle-server/pull/3379/files#diff-a524d238d651d7b718f406a41669627222a710d80a0c3267c7e0fb5c5ac4ed07
also, this is nitpicking